src/agent/rest_web_server.hpp

```diff
@@ -0,0 +1,162 @@
+}DiagInfo;
+
+typedef struct Connection {
+    
+    std::chrono::steady_clock::time_point start;
+    DiagInfo* diagInfo;
+    
+    int diagisRequest;
+    int isError;
+    int hasCallback;
+    int fd;
+    int complete;
+    int method;
+    unsigned int content_length;
+    char *path;
+    char *status;
+    char *body;
+    char *header_f;
+    char *header_v;
+    int header_state;
+    char *buf;
+    char *b;
+    int len;
+    int remain;
+    http_parser *mParser;  
+    Connection(std::chrono::steady_clock::time_point astart):start(astart),
+                                                            diagisRequest(0),
+                                                            isError(0),
+                                                            hasCallback(0),
+                                                            complete(0),
+                                                            path(NULL),
+                                                            status(NULL),
+                                                            body(NULL),
+                                                            header_f(NULL),
+                                                            header_v(NULL),
+                                                            buf(NULL),
+                                                            len(0),
+                                                            remain(8192){}
+}Connection;
+
+typedef cJSON* (*requestHandler)(Connection* connection,otInstance *  mInstance);
+typedef std::unordered_map<std::string,requestHandler> handlerMap;
+
+class RestWebServer
+{
+
+public:
+    RestWebServer(otbr::Ncp::ControllerOpenThread *aNcp);
+
+    void init();
+    
+    void UpdateFdSet(fd_set &aReadFdSet, int &aMaxFd,timeval &aTimeout);
+
+    void Process( fd_set &aReadFdSet);
+
+private:
+    /* data */
+
+    
+    void nonBlockRead(int fd);
+
+    void serveRequest(int fd);
+    
+    cJSON* getJSON(int fd);
+
+    void FreeConnection(int fd);
+
+    void parseUri(int fd);
+    
+    //for service 
+    otbr::Ncp::ControllerOpenThread *mNcp;
+    otbr::agent::ThreadHelper *   mThreadHelper ;
+    otInstance *    mInstance ;
+
+    // for server configure
+    sockaddr_in *      mAddress;
+    int mListenFd;
+
+    //for Connection
+    static const int mTimeout = 1000000;
+    int mMaxServeNum = 100;
+    std::unordered_map<int,Connection*> mConnectionSet;
+    
+    
+    // static DiagInfo* mDiagInfo;
+    
+
+    //Handler
+    handlerMap mHandlerMap;
+    void handlerInit();
+    
+    
+    static cJSON*  diagnosticRequestHandler(Connection* connection,otInstance * mInstance);
+    static void diagnosticResponseHandler(otMessage *aMessage, const otMessageInfo *aMessageInfo, void *aContext);
+    void diagnosticResponseHandler(otMessage *aMessage, const otMessageInfo &aMessageInfo);
+    
+    static cJSON*  getJsonNodeInfo(Connection* connection, otInstance * mInstance);
```

Comment:
Should those json functions go to a separate `Json` module?

Reply:
What I want to implement is that these function are `Handler` module that can be mapped to each RESTful API,  although they return pointers to  a cJSON object.  Those `GetJson` functions call a Openthread API and reformat the return value as JSON then return, so it seems better that if we stripe all codes about formatting JSON then put them into a `JSON` module then  use them after getting the result of `Handler` module.